CLDC-1687 allow deactivation or reactivation of last year schemes and locations in crossover period

arfacamble opened the pull request:
https://digital.dclg.gov.uk/jira/browse/CLDC-1687

this shows a scheme that has existed since 2020, it now cannot be deactivated before the oldest open collection period
|<img width="877" alt="image" src="https://user-images.githubusercontent.com/51094020/224697399-72406297-8fa3-4b01-a334-79d9f0914cb1.png">
|-

In this screenshot it is currently April 2023.
This scheme was started in the 23/24 collection period, so despite the other copy on the page, it cannot be deactivated before the collection period in which it was created.
|<img width="811" alt="image" src="https://user-images.githubusercontent.com/51094020/224697756-f321f2fd-e3f5-4b78-9489-26a28b502d19.png">
|-

arfacamble committed:
create method to test whether we are currently in the crossover period and associated tests

arfacamble committed:
update copy, use method for testing whether we are in the crossover period, remove focus from test file

jamesrose commented on app/helpers/collection_time_helper.rb:
```diff
@@ -44,4 +44,16 @@ def previous_collection_start_year
+      crossover_end_date += 1 until crossover_end_date.friday?
+      today.between?(Time.zone.local(today.year, 4, 1).to_date, crossover_end_date)
+    end
+  end
```
Is this the same as `FormHandler#in_crossover_period?`?

arfacamble replied:
aha. hadn't seen that.
very good.
Can I get a sense check though as this might be important for the future depending on what we're using this method for (the FormHandler one). I would expect that that method would return false for all times on the final day of the crossover period.
i.e. Time.zone.now returns the time now, but the end_date for the forms appears to be 00:00 on that day, so now < end_date will be false.
Shall I make a tech debt/bug ticket to change to now.to_date <= end_date.to_date?
I am fairly confident in that change, but I may not want to slow down a 23/24 ticket at this time writing tests.

arfacamble committed:
reuse existing method to determine whether we are in a collection period

arfacamble committed:
use the existing method and update validations

arfacamble committed:
fix test broken by changes

arfacamble committed:
update location in same way as scheme

jamesrose commented on app/models/location_deactivation_period.rb:
```diff
@@ -24,6 +26,8 @@ def validate_reactivation(record, recent_deactivation, location)
   end
 
   def validate_deactivation(record, location)
+    earliest_possible_deactivation = FormHandler.instance.in_crossover_period? ? previous_collection_start_date : current_collection_start_date
```
We're duplicating this logic four times. Could this live on the form handler itself? What would be an appropriate name to describe it?

arfacamble replied:
knock and the door shall be opened unto you...
(I agree and did this thing)

arfacamble committed:
create method on FormHandler that finds the start date of the earliest collection period

SamSeed-Softwire commented on app/models/location_deactivation_period.rb:
```diff
@@ -32,10 +34,10 @@ def validate_deactivation(record, location)
-      unless record.deactivation_date.between?(location.available_from, Time.zone.local(2200, 1, 1))
-        record.errors.add(:deactivation_date, message: I18n.t("validations.location.toggle_date.out_of_range", date: location.available_from.to_formatted_s(:govuk_date)))
-      end
+    elsif record.deactivation_date.before? FormHandler.instance.start_date_of_earliest_open_collection_period
```
Do we not want to write tests for this new case?

arfacamble replied:
I certainly considered this, but the existing validations did not yet have tests so I didn't add a whole new file.
Given we need to get this out ASAP, @jamesrose would you prefer:
- no tests
- tests just for the things I've added
- tests for all the validations here, do them now
- tests for all the validations here, raise a tech debt ticket for the next/future sprint

SamSeed-Softwire commented on app/models/scheme_deactivation_period.rb:
```diff
@@ -32,10 +34,10 @@ def validate_deactivation(record, scheme)
-      unless record.deactivation_date.between?(scheme.available_from, Time.zone.local(2200, 1, 1))
-        record.errors.add(:deactivation_date, message: I18n.t("validations.scheme.toggle_date.out_of_range", date: scheme.available_from.to_formatted_s(:govuk_date)))
-      end
+    elsif record.deactivation_date.before? FormHandler.instance.start_date_of_earliest_open_collection_period
```
Do we not want to write tests for this new case?

SamSeed-Softwire reviewed (commented):
All good except for potentially needing some tests for new error cases, I'm guessing there might be a good reason they're not required though!

arfacamble committed:
ensure that default deactivation and reactivation dates also reflect the changes

arfacamble committed:
create tests for the new validations

arfacamble committed:
lint correction

arfacamble committed:
minor copy change

arfacamble committed:
minor logic change

arfacamble committed:
amend naming error after rebase conflict
remove test that is no longer correct after change on another branch to functionality
update a test after a copy change
